New WebContent process variant for Enhanced Security

darryl-apple opened the pull request:
#### bb4d3ee9eae28a671ea11c1eae167f89412a3043
<pre>
New WebContent process variant for Enhanced Security
<a href="https://bugs.webkit.org/show_bug.cgi?id=299524">https://bugs.webkit.org/show_bug.cgi?id=299524</a>
<a href="https://rdar.apple.com/problem/161326257">rdar://problem/161326257</a>

Reviewed by Per Arne Vollan.

Implement a new WebContent Process variant for Enhanced Security

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm
* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/security/cocoa/check-enhanced-security-disabled-expected.txt: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-disabled.html: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-enabled.html: Removed.
* LayoutTests/security/cocoa/enhanced-security/check-enhanced-security-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/enhanced-security/check-enhanced-security-enabled.html: Removed.
* LayoutTests/security/cocoa/lockdown-mode/check-lockdown-mode-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/lockdown-mode/check-lockdown-mode-enabled.html: Removed.
* Source/WebCore/page/Page.h:
(WebCore::Page::setWebContentProcessVariant): Deleted.
(WebCore::Page::webContentProcessVariant const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::webContentProcessVariant const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebContentEnhancedSecurityExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.EnhancedSecurity.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-EnhancedSecurity-Info.plist: Added.
* Source/WebKit/Shared/Sandbox/iOS/webcontent-defines.sb:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _webContentProcessVariantForFrame:]):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::Client::shouldEnableEnhancedSecurity const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::webContentServiceName):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/Scripts/package-root:
* Tools/TestRunnerShared/Scripts/install-extensions-rule.sh:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnablesTrue)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnableFalse)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySamePage)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPool)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPoolReverse)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessVariantMatchesConfiguration)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessCanLaunch)):
(TestWebKitAPI::TEST(EnhancedSecurity, CaptivePortalProcessCanLaunch)):

Canonical link: <a href="https://commits.webkit.org/300869@main">https://commits.webkit.org/300869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/078a532eb5e5f4bf744fe1abcb3cb1c77389cd29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124067 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43773 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130895 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76202 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7b8e50ec-7582-4a82-84a1-3e68e6dd24ce) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52370 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94381 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62617 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/32208adf-9a99-4189-b70b-b3e621b2fb06) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127021 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110981 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74978 "Passed tests") | | ⏳ 🛠 vision-apple 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/123423 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34412 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29147 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74378 "Built successfully") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116218 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29364 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133567 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38863 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102847 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/123487 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51387 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107200 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102660 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26129 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48012 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26259 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47887 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50866 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56635 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50331 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53677 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52005 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -75,6 +75,36 @@ function mac_process_webcontent_captiveportal_entitlements()
+        fi
+    fi
+
+    mac_process_webcontent_shared_entitlements
```
Are there shared entitlements we would like to avoid having in the Enhanced Security variant?

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -394,6 +424,46 @@ function maccatalyst_process_webcontent_captiveportal_entitlements()
+    plistbuddy Add :com.apple.security.fatal-exceptions:0 string jit
+    plistbuddy Add :com.apple.developer.hardened-process bool YES
+
+    if (( "${TARGET_MAC_OS_X_VERSION_MAJOR}" >= 110000 ))
```
This version check is likely not required anymore.

darryl-apple replied:
Agreed - removed

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -394,6 +424,46 @@ function maccatalyst_process_webcontent_captiveportal_entitlements()
+        fi
+    fi
+
+    if (( "${TARGET_MAC_OS_X_VERSION_MAJOR}" >= 120000 ))
+    then
+        plistbuddy Add :com.apple.private.verified-jit bool YES
+        plistbuddy Add :com.apple.security.cs.single-jit bool YES
+    fi
```
Should these entitlements be included in the new variant?

pvollan commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -76,12 +76,25 @@
             return { "com.apple.WebKit.WebContent"_s, @"com.apple.WebKit.WebContent.CaptivePortal" };
         }
-        NSString *webContentAppex = !useCaptivePortal ? @"WebContentExtension" : @"WebContentCaptivePortalExtension";
+        NSString *webContentAppex;
```
Should we initialize this local variable to nil?

pvollan commented on Source/WebKit/Scripts/process-entitlements.sh:
```diff
@@ -265,8 +291,11 @@ function mac_process_webcontent_shared_entitlements()
 
         if (( "${TARGET_MAC_OS_X_VERSION_MAJOR}" >= 120000 ))
```
I think this version check can be removed.

pvollan commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -406,10 +422,6 @@ static ASCIILiteral serviceName(const ProcessLauncher::LaunchOptions& launchOpti
-        if (event)
-            LOG_ERROR("Error while launching %s: %s", logName.data(), xpcDictionaryGetString(event, xpcErrorDescriptionKey).utf8().data());
-        else
-            LOG_ERROR("Error while launching %s: No xpc_object_t event available.", logName.data());
```
Is this logging not necessary now?

darryl-apple replied:
Good spot. I'd removed this by accident, changed it back.

pvollan commented on Source/WebKit/Shared/AuxiliaryProcessExtensions/com.apple.WebKit.appexpt:
```diff
@@ -18,6 +18,14 @@
+        <key>EXRequiredEntitlements</key>
+        <dict>
+        </dict>
+    </dict>
```
I think this file may be unused.

darryl-apple replied:
Removed

pvollan commented on Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
```diff
@@ -24,8 +24,17 @@
+    (require-all
+        (require-not (require-entitlement "com.apple.security.cs.allow-jit"))
+        (require-not (require-entitlement "com.apple.private.verified-jit"))
+        (require-not (require-entitlement "com.apple.security.cs.single-jit"))))
```
Should we also include the private Enhanced Security entitlement in this define?

darryl-apple replied:
Good idea, changed

darryl-apple committed:
New WebContent process variant for Enhanced Security
https://bugs.webkit.org/show_bug.cgi?id=299524
rdar://problem/161326257

Reviewed by Per Arne Vollan.

Implement a new WebContent Process variant for Enhanced Security

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm
* LayoutTests/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/security/cocoa/check-enhanced-security-disabled-expected.txt: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-disabled.html: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/check-enhanced-security-enabled.html: Removed.
* LayoutTests/security/cocoa/enhanced-security/check-enhanced-security-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/enhanced-security/check-enhanced-security-enabled.html: Removed.
* LayoutTests/security/cocoa/lockdown-mode/check-lockdown-mode-enabled-expected.txt: Removed.
* LayoutTests/security/cocoa/lockdown-mode/check-lockdown-mode-enabled.html: Removed.
* Source/WebCore/page/Page.h:
(WebCore::Page::setWebContentProcessVariant): Deleted.
(WebCore::Page::webContentProcessVariant const): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::webContentProcessVariant const): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebContentEnhancedSecurityExtension.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.EnhancedSecurity.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.WebContent.sb.in:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/AuxiliaryProcessExtensions/WebContentExtension-EnhancedSecurity-Info.plist: Added.
* Source/WebKit/Shared/Sandbox/iOS/webcontent-defines.sb:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _webContentProcessVariantForFrame:]):
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::Client::shouldEnableEnhancedSecurity const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::webContentServiceName):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
* Source/WebKit/WebProcess/com.apple.WebProcess.sb.in:
* Tools/Scripts/package-root:
* Tools/TestRunnerShared/Scripts/install-extensions-rule.sh:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/EnhancedSecurity.mm:
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnablesTrue)):
(TestWebKitAPI::TEST(EnhancedSecurity, EnhancedSecurityEnableFalse)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecurity)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySamePage)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPool)):
(TestWebKitAPI::TEST(EnhancedSecurity, PSONToEnhancedSecuritySharedProcessPoolReverse)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessVariantMatchesConfiguration)):
(TestWebKitAPI::TEST(EnhancedSecurity, ProcessCanLaunch)):
(TestWebKitAPI::TEST(EnhancedSecurity, CaptivePortalProcessCanLaunch)):

Canonical link: https://commits.webkit.org/300869@main
